File: ios/MyTwitter/Models/AuthorCard.swift
import Foundation
import FirebaseFirestore

struct AuthorCard: Codable, Hashable, Identifiable {
    var id: String
    var name: String?
    var handle: String?
    var avatar: String?
    var description: String?
    var verified: Bool?
    var protected: Bool?
    var following: Bool?
    var isSelf: Bool?
    var favorited: Bool?
    var error: String?

    var displayHandle: String {
        let h = (handle ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        return h.isEmpty ? id : h
    }

    var displayName: String {
        let n = (name ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        return n.isEmpty ? "@\(displayHandle)" : n
    }
}

struct FavoriteAuthor: Codable, Hashable, Identifiable {
    @DocumentID var id: String?
    var handle: String?
    var name: String?
    var avatar: String?
    var favoritedAt: Date?
}

File: android/app/src/main/java/org/evergreenlabs/mytwitter/data/AuthorCard.kt
package org.evergreenlabs.mytwitter.data

import com.google.firebase.firestore.DocumentId
import com.google.firebase.firestore.Exclude
import com.google.firebase.firestore.IgnoreExtraProperties
import java.util.Date

@IgnoreExtraProperties
data class AuthorCard(
    val id: String = "",
    val name: String? = null,
    val handle: String? = null,
    val avatar: String? = null,
    val description: String? = null,
    val verified: Boolean? = null,
    val protected: Boolean? = null,
    val following: Boolean? = null,
    val isSelf: Boolean? = null,
    val favorited: Boolean? = null,
    val error: String? = null,
) {
    @get:Exclude
    val displayHandle: String
        get() {
            val h = handle?.trim().orEmpty()
            return h.ifEmpty { id }
        }

    @get:Exclude
    val displayName: String
        get() {
            val n = name?.trim().orEmpty()
            return n.ifEmpty { "@$displayHandle" }
        }
}

@IgnoreExtraProperties
data class FavoriteAuthor(
    @DocumentId val id: String? = null,
    val handle: String? = null,
    val name: String? = null,
    val avatar: String? = null,
    val favoritedAt: Date? = null,
)
